--- MGui.old	2012-06-20 07:39:07.000000000 +0100
+++ MGui.cpp	2012-06-20 13:37:17.555928902 +0100
@@ -28,10 +28,9 @@
 //========================================================================
 
 
-#include <MLoaders/MDevILLoader.h>
+//#include <MLoaders/MDevILLoader.h>
 #include "../Includes/MGui.h"
 
-
 // global functions
 void set2dMode(MRenderingContext * render)
 {
@@ -49,6 +48,13 @@
 	render->loadIdentity();
 }
 
+imageLoadFunc MImageLoadFunc;
+
+void setImageLoader(imageLoadFunc func)
+{
+  MImageLoadFunc = func;
+}
+
 
 // MGui
 MGui::MGui(void):
@@ -108,7 +114,7 @@
 	}
 
 	MImage image;
-	if(M_loadImage(filename, &image))
+	if(MImageLoadFunc && MImageLoadFunc(filename, &image))
 	{
 		MGuiTexture * texture = new MGuiTexture();
 
