Index: trunk/dev/MSDK/MGui/Sources/MJoystick.cpp
===================================================================
--- trunk/dev/MSDK/MGui/Sources/MJoystick.cpp	(revision 102)
+++ trunk/dev/MSDK/MGui/Sources/MJoystick.cpp	(working copy)
@@ -28,6 +28,7 @@
 //========================================================================
 
 
+#include "../Includes/MGui.h"
 #include "../Includes/MJoystick.h"
 
 
Index: trunk/dev/MSDK/MGui/Sources/MKeyboard.cpp
===================================================================
--- trunk/dev/MSDK/MGui/Sources/MKeyboard.cpp	(revision 102)
+++ trunk/dev/MSDK/MGui/Sources/MKeyboard.cpp	(working copy)
@@ -28,4 +28,5 @@
 //========================================================================
 
 
+#include "../Includes/MGui.h"
 #include "../Includes/MKeyboard.h"
\ No newline at end of file
