--- MWindow.h.orig	2012-06-20 13:43:27.000000000 +0100
+++ MWindow.h	2012-06-21 08:29:23.933978493 +0100
@@ -31,10 +31,10 @@
 #ifndef _M_WINDOW_H
 #define _M_WINDOW_H
 
-#include <MWinEvents.h>
-#include <MMouse.h>
-#include <MKeyboard.h>
-#include <MJoystick.h>
+#include "MWinEvents.h"
+#include "MMouse.h"
+#include "MKeyboard.h"
+#include "MJoystick.h"
 
 #ifdef WIN32
 	#include "WIN32/MWin32Window.h"
